src/zkevm_specs/evm/util/call_gadget.py

```diff
@@ -0,0 +1,122 @@
+from zkevm_specs.evm.table import AccountFieldTag
+from zkevm_specs.util.arithmetic import RLC
+from zkevm_specs.util.hash import EMPTY_CODE_HASH
+from zkevm_specs.util.param import (
+    GAS_COST_ACCOUNT_COLD_ACCESS,
+    GAS_COST_CALL_WITH_VALUE,
+    GAS_COST_NEW_ACCOUNT,
+    GAS_COST_WARM_ACCESS,
+)
+from ...util import (
+    FQ,
+    N_BYTES_ACCOUNT_ADDRESS,
+    N_BYTES_GAS,
+    RLC,
+)
+from ..instruction import Instruction
+
+
+class CallGadget:
+    IS_SUCCESS_CALL: FQ
+
+    gas: FQ
+    callee_address: FQ
+    value: RLC
+    cd_offset: FQ
+    cd_length: FQ
+    rd_offset: FQ
+    rd_length: FQ
+    is_success: FQ
+
+    is_u64_gas: FQ
+    next_memory_size: FQ
+    memory_expansion_gas_cost: FQ
+
+    has_value: FQ
+    callee_code_hash: FQ
+    is_empty_code_hash: FQ
+
+    def __init__(
+        self,
+        instruction: Instruction,
+        is_success_call: FQ,
+        is_call: FQ,
+        is_callcode: FQ,
+        is_delegatecall: FQ,
+    ):
+        self.instruction = instruction
+        self.IS_SUCCESS_CALL = is_success_call
+
+        # Lookup values from stack
+        self.gas_rlc = self.instruction.stack_pop()
+        callee_address_rlc = self.instruction.stack_pop()
+        # For non-OOG case,
+        # the third stack pop `value` is not present for both DELEGATECALL and
+        # STATICCALL opcodes.
+        self.value = self.instruction.stack_pop() if is_call + is_callcode == FQ(1) else RLC(0)
+        cd_offset_rlc = self.instruction.stack_pop()
+        cd_length_rlc = self.instruction.stack_pop()
+        rd_offset_rlc = self.instruction.stack_pop()
+        rd_length_rlc = self.instruction.stack_pop()
+        self.is_success = self.instruction.stack_push().expr()
+
+        if self.IS_SUCCESS_CALL == FQ(1):
+            # Verify is_success is a bool
+            self.instruction.constrain_bool(self.is_success)
+            self.gas = instruction.rlc_to_fq(self.gas_rlc, N_BYTES_GAS)
+            self.is_u64_gas = instruction.is_zero(
+                instruction.sum(self.gas_rlc.le_bytes[N_BYTES_GAS:])
+            )
+        else:
+            self.instruction.constrain_zero(self.is_success)
+            self.gas = FQ(0)
+            self.is_u64_gas = FQ(0)
```

Comment:
I don't think we need to constrain `gas` and `is_u64_gas` to zero. 

```suggestion
```

Reply:
fixed in [53df9ee](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/53df9ee479bd2b2f70207363595dafb265f91a96)